src/dataflow/src/source/mod.rs

```diff
@@ -861,6 +867,117 @@ enum MessageProcessing {
     YieldedWithDelay,
 }
 
+#[derive(Clone)]
+/// A thread-safe shareable timestamp provider that allows consumers to pause ticking of the clock
+/// while they are in the middle of producing data for a particular timestamp.
+pub struct Timestamper {
```

Comment:
This is really nice. I hope we can use something like this for all RT sources :D Just so I understand though the `Arc` is a future-proofing thing and not strictly required for the pg replica stuff today which looks like it runs on one thread right?

Reply:
Not exactly. The pg replica source is not multithreaded but because it's an async stream tokio can move the task from one thread to another, hence the requirement for Arc